server/consumer.go

```diff
@@ -2329,10 +2338,11 @@ func (o *consumer) needAck(sseq uint64) bool {
 	o.mu.RLock()
 
 	// Check first if we are filtered, and if so check if this is even applicable to us.
-	var svp StoreMsg
 	if o.isFiltered() && o.mset != nil {
+		svp := storeMsgPool.Get().(*StoreMsg)
```

Comment:
If the pooled messages hold onto interior items this could be beneficial but at first is counter intuitive, since we are adding work compared to accessing a stack variable and struct.

We could determine where we are stressed the most with the interior slices and preallocate that as well on the stack and place it into svp.

Reply:
After reviewing I can see we reuse the buf filed of a StoredMsg if present, so the benefit is coming from there for sure.

Reply:
The struct itself probably would stay on the stack in the happy case but the slices are bleeding out into the heap either way, so admittedly there's a bit of a trade-off to make here. Moving it to the heap (via the `sync.Pool`) and keeping the slice references is probably the best way to keep the underlying capacity alive, since that's where you're going to save on GC pressure. Otherwise every time the struct goes out of scope, the inner slices are marked for GC.

Am I right in thinking this code path is hot in just about any situation where `AckAll` is used? In my case I think I hit it by accident due to trying to ack messages that no longer exist (as we discussed) but it's probably also a genuine case when there *are* messages to ack too.

Reply:
I will dig in today. And thanks again for this PR and the insight, much appreciated. Will ping/tag you on PR when ready so you can review and test.